src/rt/mod.rs

```diff
@@ -149,6 +149,89 @@ where
 ///
 /// This enables concurrent algorithms that require other threads to make
 /// progress.
+///
+/// Using this as a hint might be necessary to reduce the number of branches
+/// being investigated by loom. This might be necessary when testing spin locks,
+/// since each iteration constitutes a branch point which might easily cause a
+/// combinatorical expansion.
+///
+/// Note that in loom, [`spin_loop`] and [`spin_loop_hint`] is an alias of this
+/// function.
+///
+/// [`spin_loop`]: crate::hint::spin_loop
+/// [`spin_loop_hint`]: crate::sync::atomic::spin_loop_hint
+///
+/// # Examples
+///
+/// Testing a raw spin lock under loom.
+///
+/// This is only provided as an example for when using [`spin_loop`] and
+/// [`yield_now`] could be appropriate. Using a spin lock is almost always worse
+/// than using a [`Mutex`] directly which spins internally before parking the
+/// thread if contention is detected to save on system resources.
+///
+/// [`Mutex`]: std::sync::Mutex
+/// [`spin_loop_hint`]: crate::sync::atomic::spin_loop_hint
+/// [`spin_loop`]: crate::hint::spin_loop
+///
+/// ```no_run
+/// use loom::sync::atomic::AtomicBool;
+/// use loom::hint;
+/// use loom::thread;
+///
+/// use std::sync::Arc;
+/// use std::sync::atomic::Ordering::{Acquire, Relaxed, SeqCst};
+///
+/// struct Lock {
+///     locked: AtomicBool,
+/// }
+///
+/// impl Lock {
+///     fn new() -> Self {
+///         Lock {
+///             locked: AtomicBool::new(false),
+///         }
+///     }
+///
+///     fn spin(&self) {
+///         while self.locked.load(Relaxed) {
+///             hint::spin_loop();
+///         }
+///     }
+///
+///     fn lock(&self) {
+///         loop {
+///             self.spin();
+///
+///             if self.locked.compare_exchange(false, true, Acquire, Relaxed).is_ok() {
+///                 break;
+///             }
+///
+///             thread::yield_now();
+///         }
+///     }
+///
+///     fn unlock(&self) {
+///         self.locked.store(false, SeqCst);
+///     }
+/// }
+///
+/// fn test_concurrent_logic() {
+///     loom::model(|| {
+///         let v1 = Arc::new(Lock::new());
+///         let v2 = v1.clone();
+///
+///         let t1 = thread::spawn(move || {
+///             v2.lock();
+///             // critical section.
+///             v2.unlock();
+///         });
+///
+///         v1.lock();
+///         v1.unlock();
```

Comment:
Perhaps we want to join the thread at the end?
```suggestion
///         v1.unlock();
///
///         t1.join().unwrap();
```

Reply:
I'll add it, but I also don't think it makes much of a difference.

The example code isn't testing any invariants, like correctly accessing the contents of an `UnsafeCell` and with the use of `loom::model` it should already execute all possible variants - including ones where the threads exits at the end. And here it doesn't matter since we don't have any code which depends on it.

Reply:
Perhaps I need to read a little more loom code to be sure, but yeah I also think it wouldn't make much difference here.